File: finished/subset_ii.py
from typing import List

# https://leetcode.com/problems/subsets-ii/

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        N = len(nums)
        nums = sorted(nums)

        res = []
        def helper(i: int, arr: List[int]) -> None:
            if i >= N:
                res.append(arr)
                return None
            
            # Take
            helper(i+1, arr+[nums[i]])
            # No take
            j = i + 1
            while j < N and nums[j] == nums[i]:
                j += 1
            helper(j, arr)
            return None
        helper(0, [])
        return res

File: src/go/finished/subset_ii.go
package leetcode

// https://leetcode.com/problems/subsets-ii/

import "sort"

func subsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	res := make([][]int, 0)

	var helper func([]int, int)
	helper = func(accum []int, i int) {
		if i >= len(nums) {
			cp := make([]int, len(accum))
			copy(cp, accum)
			res = append(res, cp)
			return
		}

		helper(append(accum, nums[i]), i+1)
		j := i + 1
		for j < len(nums) && nums[i] == nums[j] {
			j++
		}
		helper(accum, j)
	}

	helper([]int{}, 0)
	return res
}
